sql/rpl_trx_tracking.cc

```diff
@@ -262,15 +263,15 @@ void Writeset_trx_dependency_tracker::get_dependency(THD *thd,
     int64 last_parent = m_writeset_history_start;
     for (std::vector<uint64>::iterator it = writeset->begin();
          it != writeset->end(); ++it) {
-      Writeset_history::iterator hst = m_writeset_history.find(*it);
-      if (hst != m_writeset_history.end()) {
+      Writeset_history::iterator hst = writeset_history->find(*it);
+      if (hst != writeset_history->end()) {
         if (hst->second > last_parent && hst->second < sequence_number)
           last_parent = hst->second;
 
         hst->second = sequence_number;
       } else {
         if (!exceeds_capacity)
-          m_writeset_history.insert(
+          writeset_history->insert(
```

Comment:
It looks like we protected `m_writeset_history` shared pointer from races but what about data in the `Writeset_history` container itself?
Is `ankerl::unordered_dense::map` container thread-safe?
In other words, what happens when `get_dependency()` is called from multiple threads and each of them is doint this `insert()` operation?

Reply:
As discussed on slack, this is not an issue. The only race possible is between `get_dependency()` and `rotate()` run from multiple threads.

Reply:
As discussed over Slack, `Writeset_trx_dependency_tracker::get_dependency` is called during the binlog flush stage and it is protected with  `LOCK_log`. So there can be no concurrent insert calls to `map->insert()`

![image (4)](https://github.com/user-attachments/assets/dfd0cdfb-a292-41c1-98fc-a769761086d9)
Note that there is `mysql_mutex_assert_owner(&LOCK_log)` in line number 8786 which ensures that LOCK_log is already acquired before calling `Writeset_trx_dependency_tracker::get_dependency()`